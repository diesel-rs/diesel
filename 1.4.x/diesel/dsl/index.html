<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Includes various helper types and bare functions which are named too generically to be included in prelude, but are often used when using Diesel."><meta name="keywords" content="rust, rustlang, rust-lang, dsl"><title>diesel::dsl - Rust</title><link rel="stylesheet" type="text/css" href="../../normalize.css"><link rel="stylesheet" type="text/css" href="../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../../ayu.css" disabled ><script id="default-settings" ></script><script src="../../storage.js"></script><script src="../../crates.js"></script><noscript><link rel="stylesheet" href="../../noscript.css"></noscript><link rel="icon" type="image/svg+xml" href="../../favicon.svg"><link rel="alternate icon" type="image/png" href="../../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../../favicon-32x32.png"><style type="text/css">#crate-search{background-image:url("../../down-arrow.svg");}</style></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu" role="button">&#9776;</div><a href='../../diesel/index.html'><div class='logo-container rust-logo'><img src='../../rust-logo.png' alt='logo'></div></a><h2 class="location">Module dsl</h2><div class="sidebar-elems"><div class="block items"><ul><li><a href="#structs">Structs</a></li><li><a href="#functions">Functions</a></li><li><a href="#types">Type Definitions</a></li></ul></div><div id="sidebar-vars" data-name="dsl" data-ty="mod" data-relpath="./"></div><script defer src="./sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu" title="themes"><img width="18" height="18" alt="Pick another theme!" src="../../brush.svg"></button><div id="theme-choices" role="menu"></div></div><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input"name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><button type="button" id="help-button" title="help">?</button><a id="settings-menu" href="../../settings.html" title="settings"><img width="18" height="18" alt="Change settings" src="../../wheel.svg"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="in-band">Module <a href="../index.html">diesel</a>::<wbr><a class="mod" href="#">dsl</a><button id="copy-path" onclick="copy_path(this)" title="copy path"><img src="../../clipboard.svg" width="19" height="18" alt="Copy item import" title="Copy item import to clipboard"></button></span><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../../src/diesel/lib.rs.html#216-230" title="goto source code">[src]</a></span></h1><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Includes various helper types and bare functions which are named too
generically to be included in prelude, but are often used when using Diesel.</p>
</div></details><h2 id="reexports" class="section-header"><a href="#reexports">Re-exports</a></h2>
<div class="item-table"><div class="item-left import-item"><code>pub use pg::expression::<a class="mod" href="../pg/expression/dsl/index.html" title="mod diesel::pg::expression::dsl">dsl</a>::*;</code></div><div class="item-right docblock-short"></div></div><h2 id="structs" class="section-header"><a href="#structs">Structs</a></h2>
<div class="item-table"><div class="item-left module-item"><a class="struct" href="struct.now.html" title="diesel::dsl::now struct">now</a></div><div class="item-right docblock-short"><p>Represents the SQL <code>CURRENT_TIMESTAMP</code> constant. This is equivalent to the
<code>NOW()</code> function on backends that support it.</p>
</div></div><h2 id="functions" class="section-header"><a href="#functions">Functions</a></h2>
<div class="item-table"><div class="item-left module-item"><a class="fn" href="fn.avg.html" title="diesel::dsl::avg fn">avg</a></div><div class="item-right docblock-short"><p>Represents a SQL <code>AVG</code> function. This function can only take types which are
Foldable.</p>
</div><div class="item-left module-item"><a class="fn" href="fn.count.html" title="diesel::dsl::count fn">count</a></div><div class="item-right docblock-short"><p>Creates a SQL <code>COUNT</code> expression</p>
</div><div class="item-left module-item"><a class="fn" href="fn.count_star.html" title="diesel::dsl::count_star fn">count_star</a></div><div class="item-right docblock-short"><p>Creates a SQL <code>COUNT(*)</code> expression</p>
</div><div class="item-left module-item"><a class="fn" href="fn.date.html" title="diesel::dsl::date fn">date</a></div><div class="item-right docblock-short"><p>Represents the SQL <code>DATE</code> function. The argument should be a Timestamp
expression, and the return value will be an expression of type Date.</p>
</div><div class="item-left module-item"><a class="fn" href="fn.delete.html" title="diesel::dsl::delete fn">delete</a></div><div class="item-right docblock-short"><p>Creates a <code>DELETE</code> statement.</p>
</div><div class="item-left module-item"><a class="fn" href="fn.exists.html" title="diesel::dsl::exists fn">exists</a></div><div class="item-right docblock-short"><p>Creates a SQL <code>EXISTS</code> expression.</p>
</div><div class="item-left module-item"><a class="fn" href="fn.insert_into.html" title="diesel::dsl::insert_into fn">insert_into</a></div><div class="item-right docblock-short"><p>Creates an <code>INSERT</code> statement for the target table.</p>
</div><div class="item-left module-item"><a class="fn" href="fn.insert_or_ignore_into.html" title="diesel::dsl::insert_or_ignore_into fn">insert_or_ignore_into</a></div><div class="item-right docblock-short"><p>Creates an <code>INSERT [OR] IGNORE</code> statement.</p>
</div><div class="item-left module-item"><a class="fn" href="fn.max.html" title="diesel::dsl::max fn">max</a></div><div class="item-right docblock-short"><p>Represents a SQL <code>MAX</code> function. This function can only take types which are
ordered.</p>
</div><div class="item-left module-item"><a class="fn" href="fn.min.html" title="diesel::dsl::min fn">min</a></div><div class="item-right docblock-short"><p>Represents a SQL <code>MIN</code> function. This function can only take types which are
ordered.</p>
</div><div class="item-left module-item"><a class="fn" href="fn.not.html" title="diesel::dsl::not fn">not</a></div><div class="item-right docblock-short"><p>Creates a SQL <code>NOT</code> expression</p>
</div><div class="item-left module-item"><a class="fn" href="fn.replace_into.html" title="diesel::dsl::replace_into fn">replace_into</a></div><div class="item-right docblock-short"><p>Creates a <code>REPLACE</code> statement.</p>
</div><div class="item-left module-item"><a class="fn" href="fn.select.html" title="diesel::dsl::select fn">select</a></div><div class="item-right docblock-short"><p>Creates a bare select statement, with no from clause. Primarily used for
testing diesel itself, but likely useful for third party crates as well. The
given expressions must be selectable from anywhere.</p>
</div><div class="item-left module-item"><a class="fn" href="fn.sql.html" title="diesel::dsl::sql fn">sql</a></div><div class="item-right docblock-short"><p>Use literal SQL in the query builder</p>
</div><div class="item-left module-item"><a class="fn" href="fn.sql_query.html" title="diesel::dsl::sql_query fn">sql_query</a></div><div class="item-right docblock-short"><p>Construct a full SQL query using raw SQL.</p>
</div><div class="item-left module-item"><a class="fn" href="fn.sum.html" title="diesel::dsl::sum fn">sum</a></div><div class="item-right docblock-short"><p>Represents a SQL <code>SUM</code> function. This function can only take types which are
Foldable.</p>
</div><div class="item-left module-item"><a class="fn" href="fn.update.html" title="diesel::dsl::update fn">update</a></div><div class="item-right docblock-short"><p>Creates an <code>UPDATE</code> statement.</p>
</div></div><h2 id="types" class="section-header"><a href="#types">Type Definitions</a></h2>
<div class="item-table"><div class="item-left module-item"><a class="type" href="type.And.html" title="diesel::dsl::And type">And</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.BoolExpressionMethods.html#method.and"><code>lhs.and(rhs)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.AsExpr.html" title="diesel::dsl::AsExpr type">AsExpr</a></div><div class="item-right docblock-short"><p>The type of <code>Item</code> when converted to an expression with the same type as <code>TargetExpr</code></p>
</div><div class="item-left module-item"><a class="type" href="type.AsExprOf.html" title="diesel::dsl::AsExprOf type">AsExprOf</a></div><div class="item-right docblock-short"><p>The type of <code>Item</code> when converted to an expression of <code>Type</code></p>
</div><div class="item-left module-item"><a class="type" href="type.Asc.html" title="diesel::dsl::Asc type">Asc</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.ExpressionMethods.html#method.asc"><code>expr.asc()</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.Between.html" title="diesel::dsl::Between type">Between</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.ExpressionMethods.html#method.between"><code>lhs.between(lower, upper)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.Desc.html" title="diesel::dsl::Desc type">Desc</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.ExpressionMethods.html#method.desc"><code>expr.desc()</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.Distinct.html" title="diesel::dsl::Distinct type">Distinct</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.distinct()</code></p>
</div><div class="item-left module-item"><a class="type" href="type.DistinctOn.html" title="diesel::dsl::DistinctOn type">DistinctOn</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.distinct_on(expr)</code></p>
</div><div class="item-left module-item"><a class="type" href="type.Eq.html" title="diesel::dsl::Eq type">Eq</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.ExpressionMethods.html#method.eq"><code>lhs.eq(rhs)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.EqAny.html" title="diesel::dsl::EqAny type">EqAny</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.ExpressionMethods.html#method.eq_any"><code>lhs.eq_any(rhs)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.Escape.html" title="diesel::dsl::Escape type">Escape</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.EscapeExpressionMethods.html#method.escape"><code>lhs.escape('x')</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.Filter.html" title="diesel::dsl::Filter type">Filter</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.filter(predicate)</code></p>
</div><div class="item-left module-item"><a class="type" href="type.Find.html" title="diesel::dsl::Find type">Find</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.find(pk)</code></p>
</div><div class="item-left module-item"><a class="type" href="type.FindBy.html" title="diesel::dsl::FindBy type">FindBy</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.filter(lhs.eq(rhs))</code></p>
</div><div class="item-left module-item"><a class="type" href="type.ForKeyShare.html" title="diesel::dsl::ForKeyShare type">ForKeyShare</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.for_key_share()</code></p>
</div><div class="item-left module-item"><a class="type" href="type.ForNoKeyUpdate.html" title="diesel::dsl::ForNoKeyUpdate type">ForNoKeyUpdate</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.for_no_key_update()</code></p>
</div><div class="item-left module-item"><a class="type" href="type.ForShare.html" title="diesel::dsl::ForShare type">ForShare</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.for_share()</code></p>
</div><div class="item-left module-item"><a class="type" href="type.ForUpdate.html" title="diesel::dsl::ForUpdate type">ForUpdate</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.for_update()</code></p>
</div><div class="item-left module-item"><a class="type" href="type.Gt.html" title="diesel::dsl::Gt type">Gt</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.ExpressionMethods.html#method.gt"><code>lhs.gt(rhs)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.GtEq.html" title="diesel::dsl::GtEq type">GtEq</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.ExpressionMethods.html#method.ge"><code>lhs.ge(rhs)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.InnerJoin.html" title="diesel::dsl::InnerJoin type">InnerJoin</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.inner_join(rhs)</code></p>
</div><div class="item-left module-item"><a class="type" href="type.IntoBoxed.html" title="diesel::dsl::IntoBoxed type">IntoBoxed</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.into_boxed::&lt;'a, DB&gt;()</code></p>
</div><div class="item-left module-item"><a class="type" href="type.IsNotNull.html" title="diesel::dsl::IsNotNull type">IsNotNull</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.ExpressionMethods.html#method.is_not_null"><code>expr.is_not_null()</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.IsNull.html" title="diesel::dsl::IsNull type">IsNull</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.ExpressionMethods.html#method.is_null"><code>expr.is_null()</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.LeftJoin.html" title="diesel::dsl::LeftJoin type">LeftJoin</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.left_join(rhs)</code></p>
</div><div class="item-left module-item"><a class="type" href="type.Like.html" title="diesel::dsl::Like type">Like</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.TextExpressionMethods.html#method.like"><code>lhs.like(rhs)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.Limit.html" title="diesel::dsl::Limit type">Limit</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.limit()</code></p>
</div><div class="item-left module-item"><a class="type" href="type.Lt.html" title="diesel::dsl::Lt type">Lt</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.ExpressionMethods.html#method.lt"><code>lhs.lt(rhs)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.LtEq.html" title="diesel::dsl::LtEq type">LtEq</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.ExpressionMethods.html#method.le"><code>lhs.le(rhs)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.NeAny.html" title="diesel::dsl::NeAny type">NeAny</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.ExpressionMethods.html#method.ne_any"><code>lhs.ne_any(rhs)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.NoWait.html" title="diesel::dsl::NoWait type">NoWait</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.no_wait()</code></p>
</div><div class="item-left module-item"><a class="type" href="type.Not.html" title="diesel::dsl::Not type">Not</a><span class="stab deprecated" title="">Deprecated</span></div><div class="item-right docblock-short"><p>The return type of <code>not(expr)</code></p>
</div><div class="item-left module-item"><a class="type" href="type.NotBetween.html" title="diesel::dsl::NotBetween type">NotBetween</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.ExpressionMethods.html#method.not_between"><code>lhs.not_between(lower, upper)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.NotEq.html" title="diesel::dsl::NotEq type">NotEq</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.ExpressionMethods.html#method.ne"><code>lhs.ne(rhs)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.NotLike.html" title="diesel::dsl::NotLike type">NotLike</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.TextExpressionMethods.html#method.not_like"><code>lhs.not_like(rhs)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.Nullable.html" title="diesel::dsl::Nullable type">Nullable</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.NullableExpressionMethods.html#method.nullable"><code>expr.nullable()</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.NullableSelect.html" title="diesel::dsl::NullableSelect type">NullableSelect</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.nullable()</code></p>
</div><div class="item-left module-item"><a class="type" href="type.Offset.html" title="diesel::dsl::Offset type">Offset</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.offset()</code></p>
</div><div class="item-left module-item"><a class="type" href="type.Or.html" title="diesel::dsl::Or type">Or</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.BoolExpressionMethods.html#method.or"><code>lhs.or(rhs)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.OrFilter.html" title="diesel::dsl::OrFilter type">OrFilter</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.or_filter(predicate)</code></p>
</div><div class="item-left module-item"><a class="type" href="type.Order.html" title="diesel::dsl::Order type">Order</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.order(ordering)</code></p>
</div><div class="item-left module-item"><a class="type" href="type.Select.html" title="diesel::dsl::Select type">Select</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.select(selection)</code></p>
</div><div class="item-left module-item"><a class="type" href="type.SingleValue.html" title="diesel::dsl::SingleValue type">SingleValue</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.single_value()</code></p>
</div><div class="item-left module-item"><a class="type" href="type.SkipLocked.html" title="diesel::dsl::SkipLocked type">SkipLocked</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.skip_locked()</code></p>
</div><div class="item-left module-item"><a class="type" href="type.SqlTypeOf.html" title="diesel::dsl::SqlTypeOf type">SqlTypeOf</a></div><div class="item-right docblock-short"><p>The SQL type of an expression</p>
</div><div class="item-left module-item"><a class="type" href="type.ThenOrderBy.html" title="diesel::dsl::ThenOrderBy type">ThenOrderBy</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.then_order_by(ordering)</code></p>
</div><div class="item-left module-item"><a class="type" href="type.Update.html" title="diesel::dsl::Update type">Update</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>update(lhs).set(rhs)</code></p>
</div><div class="item-left module-item"><a class="type" href="type.avg.html" title="diesel::dsl::avg type">avg</a></div><div class="item-right docblock-short"><p>The return type of <a href="../dsl/fn.avg.html"><code>avg(expr)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.count.html" title="diesel::dsl::count type">count</a></div><div class="item-right docblock-short"><p>The return type of <a href="../dsl/fn.count.html"><code>count(expr)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.count_star.html" title="diesel::dsl::count_star type">count_star</a></div><div class="item-right docblock-short"><p>The return type of <a href="../dsl/fn.count_star.html"><code>count_star)(</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.date.html" title="diesel::dsl::date type">date</a></div><div class="item-right docblock-short"><p>The return type of <a href="../dsl/fn.date.html"><code>date(expr)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.max.html" title="diesel::dsl::max type">max</a></div><div class="item-right docblock-short"><p>The return type of <a href="../dsl/fn.max.html"><code>max(expr)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.min.html" title="diesel::dsl::min type">min</a></div><div class="item-right docblock-short"><p>The return type of <a href="../dsl/fn.min.html"><code>min(expr)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.not.html" title="diesel::dsl::not type">not</a></div><div class="item-right docblock-short"><p>The return type of <a href="../dsl/fn.not.html"><code>not(expr)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.sum.html" title="diesel::dsl::sum type">sum</a></div><div class="item-right docblock-short"><p>The return type of <a href="../dsl/fn.sum.html"><code>sum(expr)</code></a></p>
</div></div></section><section id="search" class="content hidden"></section><div id="rustdoc-vars" data-root-path="../../" data-current-crate="diesel" data-search-index-js="../../search-index.js" data-search-js="../../search.js"></div>
    <script src="../../main.js"></script>
</body></html>