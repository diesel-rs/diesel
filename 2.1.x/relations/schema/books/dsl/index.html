<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Re-exports all of the columns of this table, as well as the table struct renamed to the module name. This is meant to be glob imported for functions which only deal with one table."><title>relations::schema::books::dsl - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../../../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../../../../static.files/rustdoc-081576b923113409.css"><meta name="rustdoc-vars" data-root-path="../../../../" data-static-root-path="../../../../static.files/" data-current-crate="relations" data-themes="" data-resource-suffix="" data-rustdoc-version="1.79.0-nightly (e3181b091 2024-04-18)" data-channel="nightly" data-search-js="search-5b46ae87cc198d72.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../../../../static.files/storage-e32f0c247825364d.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../../../static.files/main-20a3ad099b048cf2.js"></script><noscript><link rel="stylesheet" href="../../../../static.files/noscript-09095024cf37855e.css"></noscript><link rel="alternate icon" type="image/png" href="../../../../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../../../../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../../../relations/index.html">relations</a><span class="version">0.1.0</span></h2></div><h2 class="location"><a href="#">Module dsl</a></h2><div class="sidebar-elems"><section><ul class="block"><li><a href="#reexports">Re-exports</a></li></ul></section><h2><a href="../index.html">In relations::schema::books</a></h2></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><div id="sidebar-button" tabindex="-1"><a href="../../../../relations/all.html" title="show sidebar"></a></div><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Type ‘S’ or ‘/’ to search, ‘?’ for more options…" type="search"><div id="help-button" tabindex="-1"><a href="../../../../help.html" title="help">?</a></div><div id="settings-menu" tabindex="-1"><a href="../../../../settings.html" title="settings">Settings</a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Module <a href="../../../index.html">relations</a>::<wbr><a href="../../index.html">schema</a>::<wbr><a href="../index.html">books</a>::<wbr><a class="mod" href="#">dsl</a><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><span class="out-of-band"><a class="src" href="../../../../src/relations/schema.rs.html#10-15">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Re-exports all of the columns of this table, as well as the
table struct renamed to the module name. This is meant to be
glob imported for functions which only deal with one table.</p>
</div></details><h2 id="reexports" class="section-header">Re-exports<a href="#reexports" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name" id="reexport.id"><code>pub use super::columns::<a class="struct" href="../columns/struct.id.html" title="struct relations::schema::books::columns::id">id</a>;</code></div></li><li><div class="item-name" id="reexport.title"><code>pub use super::columns::<a class="struct" href="../columns/struct.title.html" title="struct relations::schema::books::columns::title">title</a>;</code></div></li><li><div class="item-name" id="reexport.books"><code>pub use super::<a class="struct" href="../struct.table.html" title="struct relations::schema::books::table">table</a> as books;</code></div></li></ul></section></div></main></body></html>