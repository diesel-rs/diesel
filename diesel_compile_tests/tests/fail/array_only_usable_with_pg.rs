extern crate diesel;

use diesel::dsl::*;
use diesel::*;

fn main() {
    let mut connection = SqliteConnection::establish("").unwrap();
    select(array((1,))).get_result::<Vec<i32>>(&mut connection);

    let mut connection = MysqlConnection::establish("").unwrap();
    select(array((1,))).get_result::<Vec<i32>>(&mut connection);
}
