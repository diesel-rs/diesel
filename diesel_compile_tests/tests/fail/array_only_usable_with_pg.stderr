error[E0271]: type mismatch resolving `<diesel::SqliteConnection as diesel::Connection>::Backend == Pg`
 --> $DIR/array_only_usable_with_pg.rs:8:25
  |
8 |     select(array((1,))).get_result::<Vec<i32>>(&mut connection);
  |                         ^^^^^^^^^^ expected struct `Sqlite`, found struct `Pg`
  |
  = note: required because of the requirements on the impl of `LoadQuery<diesel::SqliteConnection, Vec<i32>>` for `SelectStatement<(), diesel::query_builder::select_clause::SelectClause<diesel::pg::expression::array::ArrayLiteral<(diesel::expression::bound::Bound<diesel::sql_types::Integer, i32>,), diesel::sql_types::Integer>>>`

error[E0277]: the trait bound `Sqlite: HasSqlType<Array<diesel::sql_types::Integer>>` is not satisfied
 --> $DIR/array_only_usable_with_pg.rs:8:25
  |
8 |     select(array((1,))).get_result::<Vec<i32>>(&mut connection);
  |                         ^^^^^^^^^^ the trait `HasSqlType<Array<diesel::sql_types::Integer>>` is not implemented for `Sqlite`
  |
  = help: the following implementations were found:
            <Sqlite as HasSqlType<BigInt>>
            <Sqlite as HasSqlType<Bool>>
            <Sqlite as HasSqlType<diesel::sql_types::Binary>>
            <Sqlite as HasSqlType<diesel::sql_types::Date>>
          and 8 others
  = note: required because of the requirements on the impl of `QueryMetadata<Array<diesel::sql_types::Integer>>` for `Sqlite`
  = note: required because of the requirements on the impl of `LoadQuery<diesel::SqliteConnection, Vec<i32>>` for `SelectStatement<(), diesel::query_builder::select_clause::SelectClause<diesel::pg::expression::array::ArrayLiteral<(diesel::expression::bound::Bound<diesel::sql_types::Integer, i32>,), diesel::sql_types::Integer>>>`

error[E0271]: type mismatch resolving `<diesel::MysqlConnection as diesel::Connection>::Backend == Pg`
  --> $DIR/array_only_usable_with_pg.rs:11:25
   |
11 |     select(array((1,))).get_result::<Vec<i32>>(&mut connection);
   |                         ^^^^^^^^^^ expected struct `Mysql`, found struct `Pg`
   |
   = note: required because of the requirements on the impl of `LoadQuery<diesel::MysqlConnection, Vec<i32>>` for `SelectStatement<(), diesel::query_builder::select_clause::SelectClause<diesel::pg::expression::array::ArrayLiteral<(diesel::expression::bound::Bound<diesel::sql_types::Integer, i32>,), diesel::sql_types::Integer>>>`

error[E0277]: the trait bound `Mysql: HasSqlType<Array<diesel::sql_types::Integer>>` is not satisfied
  --> $DIR/array_only_usable_with_pg.rs:11:25
   |
11 |     select(array((1,))).get_result::<Vec<i32>>(&mut connection);
   |                         ^^^^^^^^^^ the trait `HasSqlType<Array<diesel::sql_types::Integer>>` is not implemented for `Mysql`
   |
   = help: the following implementations were found:
             <Mysql as HasSqlType<BigInt>>
             <Mysql as HasSqlType<Bool>>
             <Mysql as HasSqlType<Datetime>>
             <Mysql as HasSqlType<Json>>
           and 15 others
   = note: required because of the requirements on the impl of `QueryMetadata<Array<diesel::sql_types::Integer>>` for `Mysql`
   = note: required because of the requirements on the impl of `LoadQuery<diesel::MysqlConnection, Vec<i32>>` for `SelectStatement<(), diesel::query_builder::select_clause::SelectClause<diesel::pg::expression::array::ArrayLiteral<(diesel::expression::bound::Bound<diesel::sql_types::Integer, i32>,), diesel::sql_types::Integer>>>`
